### ISSUE TITLE
[SMPI] replay_multiple on steroids

### ISSUE THREAD

**Issue by mpoquet** [simgrid/simgrid#129]
Hello everyone,

Sorry for the long post and big thanks to courageous readers.

Within Batsim, we would like to allow different SMPI applications to be
executed in different scenarios. Consecutively or simultaneously, the hosts
potentially being used at the same time by several MPI *executors* (different
ranks within the same application, or ranks from different applications).

We tried to implement this and greatly inspired our code from the
[SMPI replay_multiple example](https://github.com/simgrid/simgrid/blob/master/examples/smpi/replay_multiple/replay_multiple.c). However, it looks like this example and
the underlying SMPI code does not fully fit our needs :(. Notably, the way the
data is associated with each SMPI process does not seem to work in our case.

I tried to summarize our needs into a *simple* example. This example can be
found here: mpoquet/simgrid@5d73d7405 (it is not integrated into CMake yet,
it can be built by the Makefile of mpoquet/simgrid@cb123661d). To make things
easier, the example has been split into two programs later in
mpoquet/simgrid@917a98b08.

This example is mainly the same as replay_multiple, but the deployment is done
within the code rather than in XML deployment files. This example reads a
workload description file, which states (1) which jobs should be executed,
(2) at what time and (3) on which hosts. It then executes all described jobs.

Unfortunately, this simple example does not work but I tried little hacks
into SMPI to make it work. This hack can be found on mpoquet/simgrid@43c69dc808.
This hack works in the simplest case but not in more complex ones :(.
**Do you have any ideas on how to achieve what we want to do?**

# Reproducing the issue
To get the example and my dirty hack:
``` bash
git clone https://github.com/mpoquet/simgrid.git issuedir
cd issuedir
git checkout 917a98b085b
```

SimGrid must be recompiled to compile the hacky example:
``` bash
mkdir -p build
cd build
cmake ..
make
sudo make install
```

The examples are not integrated into CMake yet :
```
cd examples/smpi/replay_multiple_manual_deploy
make # or
# Uses smpi_replay_run
clang++ -g -lsimgrid -std=c++11 replay_multiple_usual.cpp -o replay_mnd_usual
# Uses smpi_replay_run_with_index (hack)
clang++ -g -lsimgrid -std=c++11 replay_multiple_hack.cpp -o replay_mnd_hack
```

# Executing jobs the usual way

If the jobs are executed via *smpi_replay_run*, the workloads seems to be
executed but the data seems to be badly associated to each SMPI process.

### Simplest workload
With the simplest workload (one job), valgrind tells that invalid data is
read/written.

``` bash
valgrind ./replay_mnd_usual workload_compute_simple \
                            ../../platforms/small_platform_with_routers.xml
```

``` valgrind
==22813== Memcheck, a memory error detector
==22813== Copyright (C) 2002-2015, and GNU GPL'd, by Julian Seward et al.
==22813== Using Valgrind-3.12.0 and LibVEX; rerun with -h for copyright info
==22813== Command: ./replay_mnd_usual workload_compute_simple ../../platforms/small_platform_with_routers.xml
==22813== 
[0.000000] [smpi_kernel/INFO] You did not set the power of the host running the simulation.  The timings will certainly not be accurate.  Use the option "--cfg=smpi/host-speed:<flops>" to set its value.Check http://simgrid.org/simgrid/latest/doc/options.html#options_smpi_bench for more information.
Job read: app='alone', file='compute_only.txt', size=2, start=0, alloc='0,1'
[Bourassa:workload_executor:(1) 0.000000] [test/INFO] Launching the job executor of job 0 (app 'alone')
[Bourassa:job_alone:(2) 0.000000] [test/INFO] Executing job 0 (smpi_app 'alone')
[Bourassa:0_0:(3) 0.000000] [test/INFO] Replaying rank 0 of job 0 (smpi_app 'alone')
==22813== Invalid write of size 4
==22813==    at 0x5165120: smpi_deployment_register_process (smpi_deployment.cpp:77)
==22813==    by 0x5165697: smpi_process_init (smpi_global.cpp:142)
==22813==    by 0x5192B7B: smpi_replay_run (smpi_replay.cpp:947)
==22813==    by 0x41603B: smpi_replay_process(int, char**) (replay_multiple.cpp:67)
==22813==    by 0x4FC16FC: simgrid::xbt::MainFunction<int (*)(int, char**)>::operator()() const (functional.hpp:48)
==22813==    by 0x4FC12CC: std::_Function_handler<void (), simgrid::xbt::MainFunction<int (*)(int, char**)> >::_M_invoke(std::_Any_data const&) (functional:1740)
==22813==    by 0x500DB9D: std::function<void ()>::operator()() const (functional:2136)
==22813==    by 0x500DB58: simgrid::kernel::context::Context::operator()() (Context.hpp:94)
==22813==    by 0x500CDDC: simgrid::kernel::context::RawContext::wrapper(void*) (ContextRaw.cpp:304)
==22813==  Address 0x85dd8b8 is 0 bytes after a block of size 8 alloc'd
==22813==    at 0x4C2AB8D: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==22813==    by 0x516559B: xbt_malloc (sysdep.h:85)
==22813==    by 0x516559B: smpi_process_init (smpi_global.cpp:130)
==22813==    by 0x5192B7B: smpi_replay_run (smpi_replay.cpp:947)
==22813==    by 0x41603B: smpi_replay_process(int, char**) (replay_multiple.cpp:67)
==22813==    by 0x4FC16FC: simgrid::xbt::MainFunction<int (*)(int, char**)>::operator()() const (functional.hpp:48)
==22813==    by 0x4FC12CC: std::_Function_handler<void (), simgrid::xbt::MainFunction<int (*)(int, char**)> >::_M_invoke(std::_Any_data const&) (functional:1740)
==22813==    by 0x500DB9D: std::function<void ()>::operator()() const (functional:2136)
==22813==    by 0x500DB58: simgrid::kernel::context::Context::operator()() (Context.hpp:94)
==22813==    by 0x500CDDC: simgrid::kernel::context::RawContext::wrapper(void*) (ContextRaw.cpp:304)
==22813== 

[... many read/write errors... ]

[Bourassa:0_0:(3) 206.219583] [test/INFO] Finished replaying rank 0 of job 0 (smpi_app 'alone')
[Fafard:0_1:(4) 206.219583] [test/INFO] Finished replaying rank 1 of job 0 (smpi_app 'alone')
[Bourassa:job_alone:(2) 206.219583] [test/INFO] Finished job 0 (smpi_app 'alone')
[206.219583] [test/INFO] Simulation finished! Final time: 206.22
==22813== 
==22813== HEAP SUMMARY:
[...]
==22813== ERROR SUMMARY: 16 errors from 8 contexts (suppressed: 0 from 0)

```

### Second simplest workload
When executing the second simplest workload (two jobs, not executed at the same
time nor on the same resources), the same kind of memory errors occur but all
jobs are executed. However, the program segfaults on SMPI_finalize and I suspect
the BoF writings to cause it.

``` bash
./replay_mnd_usual workload_compute_consecutive2 \
                   ../../platforms/small_platform_with_routers.xml
```

```
Job read: app='alone0', file='compute_only.txt', size=2, start=0, alloc='0,1'
Job read: app='alone1', file='compute_only.txt', size=2, start=1000, alloc='2,3'
[0.000000] [smpi_kernel/INFO] You did not set the power of the host running the simulation.  The timings will certainly not be accurate.  Use the option "--cfg=smpi/host-speed:<flops>" to set its value.Check http://simgrid.org/simgrid/latest/doc/options.html#options_smpi_bench for more information.
[Bourassa:workload_executor:(1) 0.000000] [test/INFO] Launching the job executor of job 0 (app 'alone0')
[Bourassa:job_alone0:(2) 0.000000] [test/INFO] Executing job 0 (smpi_app 'alone0')
[Bourassa:workload_executor:(1) 0.000000] [test/INFO] Sleeping 1000 seconds (waiting for job 1000, app 'alone1')
[Bourassa:0_0:(3) 0.000000] [test/INFO] Replaying rank 0 of job 0 (smpi_app 'alone0')
[Fafard:0_1:(4) 0.000000] [test/INFO] Replaying rank 1 of job 0 (smpi_app 'alone0')
[Bourassa:0_0:(3) 206.219583] [test/INFO] Finished replaying rank 0 of job 0 (smpi_app 'alone0')
[Fafard:0_1:(4) 206.219583] [test/INFO] Finished replaying rank 1 of job 0 (smpi_app 'alone0')
[Bourassa:job_alone0:(2) 206.219583] [test/INFO] Finished job 0 (smpi_app 'alone0')
[Bourassa:workload_executor:(1) 1000.000000] [test/INFO] Launching the job executor of job 1 (app 'alone1')
[Ginette:job_alone1:(5) 1000.000000] [test/INFO] Executing job 1 (smpi_app 'alone1')
[Ginette:1_0:(6) 1000.000000] [test/INFO] Replaying rank 0 of job 1 (smpi_app 'alone1')
[Jupiter:1_1:(7) 1000.000000] [test/INFO] Replaying rank 1 of job 1 (smpi_app 'alone1')
[Ginette:1_0:(6) 1206.219583] [test/INFO] Finished replaying rank 0 of job 1 (smpi_app 'alone1')
[Jupiter:1_1:(7) 1206.219583] [test/INFO] Finished replaying rank 1 of job 1 (smpi_app 'alone1')
[Ginette:job_alone1:(5) 1206.219583] [test/INFO] Finished job 1 (smpi_app 'alone1')
[1206.219583] [test/INFO] Simulation finished! Final time: 1206.22
*** Error in `./replay_mnd_usual': free(): invalid next size (fast): 0x00000000026b00a0 ***
[...]
```

### Analysis
When reading related parts of the SMPI code, I understood that data was
associated with each SMPI process in *process_data* global variable. The mapping
from SMPI processes to their data is done via the *index_to_process_data* global
variable.

However, I found that some assumptions were not true for this example,
notably using SIMIX process IDs as SMPI data IDs. I also tried to enable the
smpi/privatize-global-variables option, but it didn't work either and I didn't
investigate it further, as I believed it has not been made for this and would
do more complex stuff. **Should I?**

# The hack
The mpoquet/simgrid@43c69dc808 hack adds a *smpi_replay_run_with_index* function,
which allows to tell which data index (of the *process_data* global variable)
the SMPI process should use. It also adds some conditions into some functions to
use the given data index instead of the *index_to_process_data* one.

# Executing examples with the hack
When the workloads are executed with the hack, memory accesses seems fine but
the simulation does not end well with non-trivial workloads.

### Simplest workload
The simplest workload seems to work fine with the hack.

``` bash
valgrind ./replay_mnd_hack workload_compute_simple \
                           ../../platforms/small_platform_with_routers.xml
```

``` valgrind
==31139== Memcheck, a memory error detector
==31139== Copyright (C) 2002-2015, and GNU GPL'd, by Julian Seward et al.
==31139== Using Valgrind-3.12.0 and LibVEX; rerun with -h for copyright info
==31139== Command: ./replay_mnd_hack workload_compute_simple ../../platforms/small_platform_with_routers.xml
==31139== 
Job read: app='alone', file='compute_only.txt', size=2, start=0, alloc='0,1'
[0.000000] [smpi_kernel/INFO] You did not set the power of the host running the simulation.  The timings will certainly not be accurate.  Use the option "--cfg=smpi/host-speed:<flops>" to set its value.Check http://simgrid.org/simgrid/latest/doc/options.html#options_smpi_bench for more information.
[Bourassa:workload_executor:(1) 0.000000] [test/INFO] Launching the job executor of job 0 (app 'alone')
[Bourassa:job_alone:(2) 0.000000] [test/INFO] Executing job 0 (smpi_app 'alone')
[Bourassa:0_0:(3) 0.000000] [test/INFO] Replaying rank 0 of job 0 (smpi_app 'alone')
[Fafard:0_1:(4) 0.000000] [test/INFO] Replaying rank 1 of job 0 (smpi_app 'alone')
[Bourassa:0_0:(3) 206.219583] [smpi_replay/INFO] Simulation time 206.219583
[Bourassa:0_0:(3) 206.219583] [test/INFO] Finished replaying rank 0 of job 0 (smpi_app 'alone')
[Fafard:0_1:(4) 206.219583] [test/INFO] Finished replaying rank 1 of job 0 (smpi_app 'alone')
[Bourassa:job_alone:(2) 206.219583] [test/INFO] Finished job 0 (smpi_app 'alone')
[206.219583] [test/INFO] Simulation finished! Final time: 206.22
==31139== 
==31139== HEAP SUMMARY:
[...]
==31139== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 0 from 0)
```

### Second simplest workload
However, when trying the second simplest workload, it creates a
`Oops ! Deadlock or code not perfectly clean` (without any valgrind error).

``` bash
./replay_mnd_hack workload_compute_consecutive2 \
                  ../../platforms/small_platform_with_routers.xml
```

```
Job read: app='alone0', file='compute_only.txt', size=2, start=0, alloc='0,1'
Job read: app='alone1', file='compute_only.txt', size=2, start=1000, alloc='2,3'
[0.000000] [smpi_kernel/INFO] You did not set the power of the host running the simulation.  The timings will certainly not be accurate.  Use the option "--cfg=smpi/host-speed:<flops>" to set its value.Check http://simgrid.org/simgrid/latest/doc/options.html#options_smpi_bench for more information.
[Bourassa:workload_executor:(1) 0.000000] [test/INFO] Launching the job executor of job 0 (app 'alone0')
[Bourassa:job_alone0:(2) 0.000000] [test/INFO] Executing job 0 (smpi_app 'alone0')
[Bourassa:workload_executor:(1) 0.000000] [test/INFO] Sleeping 1000 seconds (waiting for job 1000, app 'alone1')
[Bourassa:0_0:(3) 0.000000] [test/INFO] Replaying rank 0 of job 0 (smpi_app 'alone0')
[Fafard:0_1:(4) 0.000000] [test/INFO] Replaying rank 1 of job 0 (smpi_app 'alone0')
[Bourassa:0_0:(3) 206.219583] [test/INFO] Finished replaying rank 0 of job 0 (smpi_app 'alone0')
[Fafard:0_1:(4) 206.219583] [test/INFO] Finished replaying rank 1 of job 0 (smpi_app 'alone0')
[Bourassa:job_alone0:(2) 206.219583] [test/INFO] Finished job 0 (smpi_app 'alone0')
[Bourassa:workload_executor:(1) 1000.000000] [test/INFO] Launching the job executor of job 1 (app 'alone1')
[Ginette:job_alone1:(5) 1000.000000] [test/INFO] Executing job 1 (smpi_app 'alone1')
[Ginette:1_0:(6) 1000.000000] [test/INFO] Replaying rank 0 of job 1 (smpi_app 'alone1')
[Jupiter:1_1:(7) 1000.000000] [test/INFO] Replaying rank 1 of job 1 (smpi_app 'alone1')
[1206.219583] /home/carni/proj/simgrid-martin/src/simix/smx_global.cpp:558: [simix_kernel/CRITICAL] Oops ! Deadlock or code not perfectly clean.
[1206.219583] [simix_kernel/INFO] 3 processes are still running, waiting for something.
[1206.219583] [simix_kernel/INFO] Legend of the following listing: "Process <pid> (<name>@<host>): <status>"
[1206.219583] [simix_kernel/INFO] Process 5 (job_alone1@Ginette): waiting for synchronization synchro 0x25366e0 () in state 0 to finish
[1206.219583] [simix_kernel/INFO] Process 6 (1_0@Ginette): waiting for synchronization synchro 0x2536740 () in state 0 to finish
[1206.219583] [simix_kernel/INFO] Process 7 (1_1@Jupiter): waiting for synchronization synchro 0x2482b60 () in state 0 to finish
Abandon (core dumped)
```

### Analysis
**I don't know where this deadlock comes from, do you have any hints?**

**Comment by mquinson** [comment]
Hello,

I must confess I'm not very keen of your hack. I mean that I'd prefer not to integrate this code in the main tree of simgrid. That's OK if you don't plan to integrate it anyway, but then it seems dangerous to see your branch diverging too much from the main branch. I came to use http://savannah.nongnu.org/projects/quilt/ rather heavily during my own PhD, but I'd say that SimGrid is a moving target that will make things really unpleasant for you if you go too much in this direction.

My preferred solution would be if you could investigate a bit on why the assumption of the SMPI/SimDag bindings are not holding in this case. IIRC, they are the fruit of a later change, as initially, SimDag and SMPI were not supposed to run together. IIRC, Luka was the one who did it. Your problem may be the result of something overseen during the bindings of the both. But I'd understand if you don't want to spend too much time on improving this part of the code.

And finally, I feel very unconfortable here because I never really digged into the SMPI code so I cannot really help you there. I guess that Augustin or Luka could help, maybe.

**Comment by mpoquet** [comment]
Thank you for your answer.

I completely agree with this patch not being integrated, it is quite dirty and
was only done to investigate the memory issues I got in Batsim.

In the long run I would like to:
  - Improve the SMPI code so that *smpi_replay_run* would also work in the new
    case (probably by changing some internal data structures, not with a plenty
    of *if* nor by forcing the user to map the data manually)
  - Integrate the *usual* example into the test system

I think the current code was done with a few assumptions about the replay
context. For example, I suspect that the number of non-SMPI processes was
supposed to be 1 for the whole SMPI simulation ; but it varies dynamically
in the example case and cannot be predicted in the Batsim case. A few tests
show this behaviour below.

I think I will try to replace the **process_data** and **index_to_process_data**
global fixed-size arrays with a more dynamic structure like a global std::map
associating SMPI instances to their data. It might allow to:
  - Avoid the current invalid memory accesses in the new cases,
  - Keep only one *smpi_replay_run* entry function. As the function knows the
    application name and the process rank, it should be enough to retrieve the
    associated data,
  - Improve dynamicity, as applications would no longer need to be
    registered at the beginning of the simulation

I don't think it will fix the deadlock issue, but as it doesn't happen in
currently working examples, it would not be an immediate problem to integrate
this modification into SimGrid.
Please let me know if you see any problem with a global std::map approach, or
if you have a better idea to improve the data structure.
I will also try to merge it with the way the data is mapped when privatization
is enabled, but I will first try to understand the whole code behind it.

### Simple run of the simple usual example
When the *usual* example is run directly (as if without any hacky commit),
the data mapping is done like this:
  - **process_data** is initialised with **process_count** = 2 (perfectly fine)
  - **index_to_process_data** is initialised with *SIMIX_process_count()* = 2 (why not)
  - The index of rank 0 of the only app is **2**, which is bad as it is directly
    used to access the array **index_to_process_data**, whose size is **2**.
  - The index of rank 1 of the only app is **3**, which is bad for the same
    reasons.

``` bash
gdb -q --args ./replay_mnd_usual workload_compute_simple \
                                 ../../platforms/small_platform_with_routers.xml
```

``` gdb
[...]


(gdb) start
[...]


(gdb) break smpi_global.cpp:130
Breakpoint 2 at 0x7ffff7a50304: file /home/carni/proj/simgrid/src/smpi/smpi_global.cpp, line 130.
(gdb) break smpi_global.cpp:142
Breakpoint 3 at 0x7ffff7a50400: file /home/carni/proj/simgrid/src/smpi/smpi_global.cpp, line 142.
(gdb) break smpi_global.cpp:605
Breakpoint 4 at 0x7ffff7a51b79: file /home/carni/proj/simgrid/src/smpi/smpi_global.cpp, line 605.


(gdb) continue
Continuing.
Job read: app='alone', file='compute_only.txt', size=2, start=0, alloc='0,1'
Breakpoint 4, smpi_global_init () at /home/carni/proj/simgrid/src/smpi/smpi_global.cpp:605
605   process_data       = new smpi_process_data_t[process_count];
(gdb) p process_count
$1 = 2


(gdb) continue
Continuing.
[0.000000] [smpi_kernel/INFO] You did not set the power of the host running the simulation.  The timings will certainly not be accurate.  Use the option "--cfg=smpi/host-speed:<flops>" to set its value.Check http://simgrid.org/simgrid/latest/doc/options.html#options_smpi_bench for more information.
[Bourassa:workload_executor:(1) 0.000000] [test/INFO] Launching the job executor of job 0 (app 'alone')
[Bourassa:job_alone:(2) 0.000000] [test/INFO] Executing job 0 (smpi_app 'alone')
[Bourassa:0_0:(3) 0.000000] [test/INFO] Replaying rank 0 of job 0 (smpi_app 'alone')
Breakpoint 2, smpi_process_init (argc=0x7ffff4a2cebc, argv=0x7ffff4a2ceb0) at /home/carni/proj/simgrid/src/smpi/smpi_global.cpp:130
130       index_to_process_data=static_cast<int*>(xbt_malloc(map_size*sizeof(int)));
(gdb) p SIMIX_process_count()
$2 = 2
(gdb) p map_size
$3 = 2



(gdb) continue
Continuing.
Breakpoint 3, smpi_process_init (argc=0x7ffff4a2cebc, argv=0x7ffff4a2ceb0) at /home/carni/proj/simgrid/src/smpi/smpi_global.cpp:142
142     smpi_deployment_register_process(instance_id, rank, index, &temp_comm_world, &temp_bar);
(gdb) p rank
$4 = 0
(gdb) p index
$5 = 2


(gdb) continue
Continuing.
[Fafard:0_1:(4) 0.000000] [test/INFO] Replaying rank 1 of job 0 (smpi_app 'alone')
Breakpoint 3, smpi_process_init (argc=0x7ffff4229ebc, argv=0x7ffff4229eb0) at /home/carni/proj/simgrid/src/smpi/smpi_global.cpp:142
142     smpi_deployment_register_process(instance_id, rank, index, &temp_comm_world, &temp_bar);
(gdb) p rank
$6 = 1
(gdb) p index
$7 = 3
```

### Adding initial noise processes

Let's add some MSG processes at the beginning of the simulation to see what
happens. It can be done by removing the comments in the *main* function of
replay_multiple_usual.cpp :
``` C++
pop_some_processes(3000, hosts[0]);
```

Recompiling the example:
``` bash
make
```

Running it with gdb gives those results:
  - **process_data** is initialised with **process_count** = 2 (perfectly fine)
  - **index_to_process_data** is initialised with *SIMIX_process_count()* = 3002 (why not, but uses too much memory)
  - The index of rank 0 of the only app is **3002**, which is bad as it is directly
    used to access the array **index_to_process_data**, whose size is **3002**.
  - The index of rank 1 of the only app is **3003**, which is bad for the same
    reasons.

### Adding non-initial noise processes
In the real Batsim case, noise processes are executed between the jobs
executions, which increase even more the process IDs. In this case, the whole
idea behind the fixed-size **index_to_process_data** may be wrong, as it should
be really big to handle all cases. Using a dynamic array for
**index_to_process_data** may fix the problem, but using a map instead might
reduce memory wastes and make the problem easier.

**Comment by mquinson** [comment]
Hello,

your proposal seem very reasonable. Did you do any progress on its implementation? Do you need help to add the "usual example" as a regular test of SimGrid? 

Thanks for your work,
Mt

**Comment by mpoquet** [comment]
Hello,

Unfortunately I've been sick the entire holiday period and I didn't work on it :(.

This afternoon, @alegrand helped us to understand the different SMPI use cases and we looked at the SMPI code together. I think I'm quite fine with the code now and I'll try to implement the proposal tomorrow.

Is [this page](http://simgrid.gforge.inria.fr/simgrid/lastest/doc/inside_tests.html#inside_tests_add_integration) up-to-date? If so, I think I'll be able to add the *usual example* as a regular test :).

**Comment by mquinson** [comment]
I'm sorry to ear that you were sick. There is no hurry on my side.
The page you're pointing to is mostly uptodate. I'll add a word about the importance of documenting your code and making tests that are easy to diagnose when they break.

Thanks for your work and have a good one.
Mt

Le 4 janvier 2017 18:19:16 GMT+01:00, mpoquet <notifications@github.com> a écrit :
>Hello,
>
>Unfortunately I've been sick the entire holiday period and I didn't
>work on it :(.
>
>This afternoon, @alegrand helped us to understand the different SMPI
>use cases and we looked at the SMPI code together. I think I'm quite
>fine with the code now and I'll try to implement the proposal tomorrow.
>
>Is [this
>page](http://simgrid.gforge.inria.fr/simgrid/lastest/doc/inside_tests.html#inside_tests_add_integration)
>up-to-date? If so, I think I'll be able to add the *usual example* as a
>regular test :).
>
>
>-- 
>You are receiving this because you commented.
>Reply to this email directly or view it on GitHub:
>https://github.com/simgrid/simgrid/issues/129#issuecomment-270429097

-- 
Envoyé de mon téléphone. Veuillez excuser ma brièveté.

**Comment by mpoquet** [comment]
Hello,

Today I implemented in mpoquet/simgrid@d3ce0f03b36f the:
  - removal of the **index_to_process_data** array
  - removal of the **process_data** array
  - addition of the **smpi_processes_data** map, which associates SMPI indexes with their data

Something is obviously wrong about it since most tests are now broken.
However, this commit allows the *usual_example* to work on simple workloads (without communications) which was not the case before.

### Works on non-trivial workloads WITHOUT communications
``` bash
valgrind ./replay_mnd_usual ../../platforms/small_platform_with_routers.xml workload_compute
```

``` valgrind
==14664== Memcheck, a memory error detector
==14664== Copyright (C) 2002-2015, and GNU GPL'd, by Julian Seward et al.
==14664== Using Valgrind-3.12.0 and LibVEX; rerun with -h for copyright info
==14664== Command: ./replay_mnd_usual ../../platforms/small_platform_with_routers.xml workload_compute
==14664== 
Job read: app='alone', file='compute_only.txt', size=2, start=0, alloc='0,1'
Job read: app='not_alone0', file='compute_only.txt', size=2, start=1000, alloc='0,1'
Job read: app='not_alone1', file='compute_only.txt', size=2, start=1001, alloc='2,3'
Job read: app='time_sharing0', file='compute_only.txt', size=2, start=2000, alloc='0,1'
Job read: app='time_sharing1', file='compute_only.txt', size=2, start=2001, alloc='0,1'
[0.000000] [smpi_kernel/INFO] You did not set the power of the host running the simulation.  The timings will certainly not be accurate.  Use the option "--cfg=smpi/host-speed:<flops>" to set its value.Check http://simgrid.org/simgrid/latest/doc/options.html#options_smpi_bench for more information.
[Bourassa:workload_executor:(1) 0.000000] [test/INFO] Launching the job executor of job 0 (app 'alone')
[Bourassa:job_alone:(2) 0.000000] [test/INFO] Executing job 0 (smpi_app 'alone')
[Bourassa:workload_executor:(1) 0.000000] [test/INFO] Sleeping 1000 seconds (waiting for job 1000, app 'not_alone0')
[Bourassa:0_0:(3) 0.000000] [test/INFO] Replaying rank 0 of job 0 (smpi_app 'alone')
[Fafard:0_1:(4) 0.000000] [test/INFO] Replaying rank 1 of job 0 (smpi_app 'alone')
[Bourassa:0_0:(3) 206.219583] [test/INFO] Finished replaying rank 0 of job 0 (smpi_app 'alone')
[Fafard:0_1:(4) 206.219583] [test/INFO] Finished replaying rank 1 of job 0 (smpi_app 'alone')
[Bourassa:job_alone:(2) 206.219583] [test/INFO] Finished job 0 (smpi_app 'alone')
[Bourassa:workload_executor:(1) 1000.000000] [test/INFO] Launching the job executor of job 1 (app 'not_alone0')
[Bourassa:job_not_alone0:(5) 1000.000000] [test/INFO] Executing job 1 (smpi_app 'not_alone0')
[Bourassa:workload_executor:(1) 1000.000000] [test/INFO] Sleeping 1 seconds (waiting for job 1001, app 'not_alone1')
[Bourassa:1_0:(6) 1000.000000] [test/INFO] Replaying rank 0 of job 1 (smpi_app 'not_alone0')
[Fafard:1_1:(7) 1000.000000] [test/INFO] Replaying rank 1 of job 1 (smpi_app 'not_alone0')
[Bourassa:workload_executor:(1) 1001.000000] [test/INFO] Launching the job executor of job 2 (app 'not_alone1')
[Ginette:job_not_alone1:(8) 1001.000000] [test/INFO] Executing job 2 (smpi_app 'not_alone1')
[Bourassa:workload_executor:(1) 1001.000000] [test/INFO] Sleeping 999 seconds (waiting for job 2000, app 'time_sharing0')
[Ginette:2_0:(9) 1001.000000] [test/INFO] Replaying rank 0 of job 2 (smpi_app 'not_alone1')
[Jupiter:2_1:(10) 1001.000000] [test/INFO] Replaying rank 1 of job 2 (smpi_app 'not_alone1')
[Bourassa:1_0:(6) 1206.219583] [test/INFO] Finished replaying rank 0 of job 1 (smpi_app 'not_alone0')
[Fafard:1_1:(7) 1206.219583] [test/INFO] Finished replaying rank 1 of job 1 (smpi_app 'not_alone0')
[Bourassa:job_not_alone0:(5) 1206.219583] [test/INFO] Finished job 1 (smpi_app 'not_alone0')
[Ginette:2_0:(9) 1207.219583] [test/INFO] Finished replaying rank 0 of job 2 (smpi_app 'not_alone1')
[Jupiter:2_1:(10) 1207.219583] [test/INFO] Finished replaying rank 1 of job 2 (smpi_app 'not_alone1')
[Ginette:job_not_alone1:(8) 1207.219583] [test/INFO] Finished job 2 (smpi_app 'not_alone1')
[Bourassa:workload_executor:(1) 2000.000000] [test/INFO] Launching the job executor of job 3 (app 'time_sharing0')
[Bourassa:job_time_sharing0:(11) 2000.000000] [test/INFO] Executing job 3 (smpi_app 'time_sharing0')
[Bourassa:workload_executor:(1) 2000.000000] [test/INFO] Sleeping 1 seconds (waiting for job 2001, app 'time_sharing1')
[Bourassa:3_0:(12) 2000.000000] [test/INFO] Replaying rank 0 of job 3 (smpi_app 'time_sharing0')
[Fafard:3_1:(13) 2000.000000] [test/INFO] Replaying rank 1 of job 3 (smpi_app 'time_sharing0')
[Bourassa:workload_executor:(1) 2001.000000] [test/INFO] Launching the job executor of job 4 (app 'time_sharing1')
[Bourassa:job_time_sharing1:(14) 2001.000000] [test/INFO] Executing job 4 (smpi_app 'time_sharing1')
[Bourassa:4_0:(15) 2001.000000] [test/INFO] Replaying rank 0 of job 4 (smpi_app 'time_sharing1')
[Fafard:4_1:(16) 2001.000000] [test/INFO] Replaying rank 1 of job 4 (smpi_app 'time_sharing1')
[Bourassa:3_0:(12) 2411.439165] [test/INFO] Finished replaying rank 0 of job 3 (smpi_app 'time_sharing0')
[Fafard:3_1:(13) 2411.439165] [test/INFO] Finished replaying rank 1 of job 3 (smpi_app 'time_sharing0')
[Bourassa:job_time_sharing0:(11) 2411.439165] [test/INFO] Finished job 3 (smpi_app 'time_sharing0')
[Bourassa:4_0:(15) 2412.439165] [test/INFO] Finished replaying rank 0 of job 4 (smpi_app 'time_sharing1')
[Fafard:4_1:(16) 2412.439165] [test/INFO] Finished replaying rank 1 of job 4 (smpi_app 'time_sharing1')
[Bourassa:job_time_sharing1:(14) 2412.439165] [test/INFO] Finished job 4 (smpi_app 'time_sharing1')
[2412.439165] [test/INFO] Simulation finished! Final time: 2412.44
==14664== 
==14664== HEAP SUMMARY:
[...]
```

### Crashes on trivial workloads (1 job) WITH communication
``` bash
gdb -q --args ./replay_mnd_usual ../../platforms/small_platform_with_routers.xml workload_mixed1
```

``` gdb
Reading symbols from ./replay_mnd_usual...done.
(gdb) run
Starting program: /home/carni/proj/simgrid-mpoquet/examples/smpi/replay_multiple_manual_deploy/replay_mnd_usual ../../platforms/small_platform_with_routers.xml workload_mixed1
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/usr/lib/libthread_db.so.1".
Job read: app='alone', file='mixed.txt', size=2, start=0, alloc='0,1'
[0.000000] [smpi_kernel/INFO] You did not set the power of the host running the simulation.  The timings will certainly not be accurate.  Use the option "--cfg=smpi/host-speed:<flops>" to set its value.Check http://simgrid.org/simgrid/latest/doc/options.html#options_smpi_bench for more information.
[Bourassa:workload_executor:(1) 0.000000] [test/INFO] Launching the job executor of job 0 (app 'alone')
[Bourassa:job_alone:(2) 0.000000] [test/INFO] Executing job 0 (smpi_app 'alone')
[Bourassa:0_0:(3) 0.000000] [test/INFO] Replaying rank 0 of job 0 (smpi_app 'alone')
[Fafard:0_1:(4) 0.000000] [test/INFO] Replaying rank 1 of job 0 (smpi_app 'alone')

Program received signal SIGSEGV, Segmentation fault.
0x00007ffff7934c94 in simcall_HANDLER_comm_isend (simcall=0x6896e0, src_proc=0x0, mbox=0x688af0, task_size=1000000000, rate=-1, src_buff=0x0, src_buff_size=1000000000, match_fun=0x7ffff7a461be <match_send(void*, void*, smx_activity_t)>, 
    clean_fun=0x7ffff7a03a82 <xbt_free_f(void*)>, copy_data_fun=0x7ffff7a5ee26 <smpi_comm_null_copy_buffer_callback(smx_activity_t, void*, size_t)>, data=0x689420, detached=0)
    at /home/carni/proj/simgrid-mpoquet/src/simix/smx_network.cpp:211
211	  xbt_fifo_push(src_proc->comms, other_synchro);
(gdb) bt
#0  0x00007ffff7934c94 in simcall_HANDLER_comm_isend (simcall=0x6896e0, src_proc=0x0, mbox=0x688af0, task_size=1000000000, rate=-1, src_buff=0x0, src_buff_size=1000000000, 
    match_fun=0x7ffff7a461be <match_send(void*, void*, smx_activity_t)>, clean_fun=0x7ffff7a03a82 <xbt_free_f(void*)>, copy_data_fun=0x7ffff7a5ee26 <smpi_comm_null_copy_buffer_callback(smx_activity_t, void*, size_t)>, data=0x689420, 
    detached=0) at /home/carni/proj/simgrid-mpoquet/src/simix/smx_network.cpp:211
#1  0x00007ffff7946a6a in SIMIX_simcall_handle (simcall=0x6896e0, value=0) at /home/carni/proj/simgrid-mpoquet/src/simix/popping_generated.cpp:230
#2  0x00007ffff792abf8 in SIMIX_run () at /home/carni/proj/simgrid-mpoquet/src/simix/smx_global.cpp:510
#3  0x00007ffff78d4e66 in MSG_main () at /home/carni/proj/simgrid-mpoquet/src/msg/msg_global.cpp:96
#4  0x0000000000417738 in main (argc=3, argv=0x7fffffffde68) at replay_multiple_usual.cpp:320
(gdb) p src_proc
$1 = (smx_actor_t) 0x0
```

I tried to investigate the problem but I have no idea why src_proc is null.
Do you have any idea on what I forgot?

**Comment by mpoquet** [comment]
I worked on it with @alegrand and the **smpi_processes_data** global map now uses simix process IDs as keys.
Our test case now (mpoquet/simgrid@587483ebe7) works with all input workloads!
However, most smpi tests still break. I will try to fix it.

**Comment by mquinson** [comment]
On Fri, Jan 06, 2017 at 06:51:02AM -0800, mpoquet wrote:
> I worked on it with @alegrand and the **smpi_processes_data** global map now uses simix process IDs as keys.
> Our test case now (mpoquet/simgrid@587483ebe7) works with all input workloads!
> However, most smpi tests still break. I will try to fix it.


That's a great news! Kudos. I wish I could help you. But it's not
quite the case unfortunately....

**Comment by mquinson** [comment]
Actually not. I have the feeling that #176 was closed because it is superseeded by #271, but this bug report should not be closed without the approval of @mpoquet. Thus reopening.

**Comment by mpoquet** [comment]
PR #271 adds support in the SG code for this feature.
I'll merge the replay on steroids examples later on (will do another PR just after #271 's merge).

**Comment by mquinson** [comment]
What is the status of this? I seem to understand tat #280 superseeds it and that we should close it. @mpoquet you confirm?

**Comment by mpoquet** [comment]
I confirm that #271 and #280 dealt with this issue.
